### PR TITLE
Prevent original document instances from modifications

### PR DESCRIPTION
**Pull request by DanStory**
- Fix document references with clone (by default) before any modifications happen.
- Kept old behavior of `save` updating original reference but fixing order after post events.
- Provide a way to revert back to original behavior of events via collection opts or passed in singularly via `invokeEvents`
- Updated Mongo pkg to latest.
- Updated to latest Swimlane linting rules.

- Version bump to v5.0.0

### CODE REVIEW

File: tslint.json [hunk 1]
```diff
@@ -1,3 +1,6 @@
 {
-  "extends": "tslint-config-swimlane/base"
+  "extends": "tslint-config-swimlane/base",
+  "rules": {
+    "no-any": false
```
Comment (shaunburdick) [inline tslint.json:4]: 🤬

File: src/Repository.ts [hunk 2]
```diff
@@ -301,16 +305,28 @@ export class MongoRepository<DOC, DTO = DOC> {
...
   protected async invokeEvents(
     type: string,
     fns: RepoOperation[],
     newDocument: any,
-    originalDocument?: any
+    originalDocument?: any,
+    opts?: EventOptions
   ): Promise<any> {
+    // local options override global collection options
+    if (opts?.noClone === false || (opts?.noClone === undefined && this.options.eventOpts?.noClone !== true)) {
```
Comment (shaunburdick) [inline src/Repository.ts:321]: Does this turn on cloning by default? If so, then maybe this should be a major semver bump?

File: tslint.json (hunk 1 above)
Comment (DanStory) [reply tslint.json:4]: // TODO: resolve all the anys

File: src/Repository.ts (hunk 2 above)
Comment (DanStory) [reply src/Repository.ts:321]: it does 👍

Comment (DanStory) [reply src/Repository.ts:321]: Fixed.

Comment (shaunburdick) [reply src/Repository.ts:321]: 😇

File: tslint.json (hunk 1 above)
Comment (DanStory) [reply tslint.json:4]: https://github.com/swimlane/mongtype/issues/29#issue-810174793

**Review by sjgrannan — APPROVED** [review]

**Pull request merged**